<h1 id="fg_regulatory_features">Regulatory Build</h1>
<p>
The regulatory build provides a single "best guess" set of regulatory elements.  These elements are based on the information contained within the ensembl-functional genomics ('funcgen') database.
</p>
<p>
Regulatory features are built as a composite set of annotations based on co-occurence analysis and classification of multiple genome wide epigenomic data sets.  Each feature was built on some or all of the following data sets:
</p>

<table border="1">
<tr>
<th ALIGN='LEFT'>Anchor/Focus Sets</th>
<th>Data type</th>
<th>Reference</th>
</tr>
<tr>
<td>DNase1 Hypersensitivity site (GM06990 and CD4+ T cells) </td>
<td>ChIP-Seq</td>
<td>1</td>
</tr>
<tr>
<td>CCCTC-binding factor (CTCF)</td>
<td>ChIP-Chip*</td>
<td>2</td>
</tr>
<tr>
<td>CCTC-binding factor (CTCF)</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>
</table>
<p>
</p>
<table border="1">
<tr>
<th>Supporting Sets</th>
<th>Data type</th>
<th>Reference</th>
</tr>
<tr>
<td>H4K20me3</td>
<td>ChIP-Chip</td>
<td>3</td>
</tr>
<tr>
<td>H3K27me3</td>
<td>ChIP-Chip</td>
<td>3</d>
<tr>
<tr>
<td>H3K36me3</td>								<td>ChIP-Chip</td>
<td>3</td>
</tr>
<tr>
<td>H3K79me3</td>									<td>ChIP-Chip</td>
<td>3</td>
</tr>
<tr>
<td>H3K9me3</td>									<td>ChIP-Chip</td>
<td>3</td>
</tr>

<tr>
<td>H2AZ</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H2BK5me1</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3K27me1, me2, me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3K36me1, me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3K4me1,me2,me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3K79me1,me2,me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3K9me1,me2,me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H3R2me1, me2</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H4K20me1, me3</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>H4R3me2</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

<tr>
<td>PolII</td>
<td>ChIP-Seq</td>
<td>4</td>
</tr>

</table>
<p>
The anchor/focus sets were chosen to define a set of regions as potentially regulatory and for their previously known specific properties including DNaseI as a marker of open chromatin, H4K4me3 association with active promoters, and CTCF's association with "insulator regions."
</p>
<p>
In short, the Regulatory Build process performs an overlap analysis on each anchor/focus set with respect to each other and each of the supporting sets.  The result of this analysis was then combined into one 'RegulatoryFeature' set, merging constituent feature boundaries up to a maximium of 4KB and integrating information on proximity (<2.5KB) to transcription start and end sites.  The 4KB limit was chosen to avoid chaining of large regulatory regions (e.g the Hox cluster) in an attempt to provide more granularity over regions of interest.  On breach of the 4KB maximum length limit, features were broken down into H3K4me3 elements, or CTCF elements where appropriate.
</p>
<p>
The composite regulatory features were then classified by the patterns of data sets observed across each regulatory feature.  Some preliminary analysis has identified the following combinations that are common and strongly associated with other annotated features in Ensembl:
</p>
<ul>
<li>non-gene-associated = patterns under-represented in genic regions</li>
<li>gene-associated = patterns over-represented in genic regions</li>

<li>promoter-associated = patterns over-represented in the region of the first exon and 2500bp upstream of protein coding genes, but not in the downstream 'gene-body'. </li>
</ul>
<p>
We have used these patterns of the basis of annotations used in this version of the Ensembl Regulatory Build.
</p>
<p>
<p>
These data sets can be displayed along the chromosome in <a href="/Help/View?id=140" class="cp-internal">'Region in Detail'</a>, displayed for a gene in the <a href="/Help/View?id=142" class="cp-internal">'Regulation' view</a>or mined from the <a href="/info/docs/api/index.html" class="cp-external">functional genomics database</a>.
<p>
<strong>References</strong>
</p>
<p>
1.  Genome-wide identification of DNaseI hypersensitive sites was performed by Greg Crawford and Terry Furey (Duke University) using a whole genome DNase-sequencing protocol (Crawford et al., Genome Research 2006).<br> DNase-sequencing was performed using the Illumina (Solexa) sequencing by synthesis method from a DNase treated library generated from the GM06990 cell line (Crawford and Furey, unpublished).
</p>
<p>
2.  Kim, T.H.; Abdullaev, Z.K.; Smith, A.D.; Ching, K.A.; Loukinov, D.I.; Green, R.D.; Zhang, M.Q.; Lobanenkov, V.V. & Ren, B.<br>
Analysis of the vertebrate insulator protein CTCF-binding sites in the human genome.<br>
Cell, 2007 , 128 , 1231-1245
</p>
<p>
3.  Hirst, M; Hurd, P.J.; Bainbridge, M.; Robertson, G.; Kirmizis, A.; Nelson, C.; Zhao, Y.; Zeng, T.; Pandoh, P.; Tam, A.; Prabhu, A.; Dhalla, N.; Sa, D.; Delaney, A.; Bilenky, M.; Jones, S.; Kouzarides, T.; Marra, M.  (In preparation)
</p>
<p>
4. A. Barski, S. Cuddapah, K. Cui, T.Y. Roh, D.E. Schones, Z. Wang, G. Wei, I. Chepelev and K. Zhao, (2007). High-resolution profiling of histone methylations in the human genome, Cell 129 (2007), pp. 823-837.
</p>
<p>
* The CTCF data was processed with the Nessie HMM (Flicek, unpublished).
</p>
